src/interface/contract.rs

```diff
@@ -155,21 +156,23 @@ impl StateChange for OwnedFractionChange {
         *self = match self {
             OwnedFractionChange::Dec(neg) => OwnedFractionChange::Dec(*neg + sub),
             OwnedFractionChange::Zero => OwnedFractionChange::Dec(sub),
-            OwnedFractionChange::Inc(pos) if *pos > sub => OwnedFractionChange::Inc(*pos - sub),
-            OwnedFractionChange::Inc(pos) if *pos == sub => OwnedFractionChange::Zero,
-            OwnedFractionChange::Inc(pos) if *pos < sub => OwnedFractionChange::Dec(sub - *pos),
-            OwnedFractionChange::Inc(_) => unreachable!(),
+            OwnedFractionChange::Inc(pos) => match sub.cmp(pos) {
+                Ordering::Less => OwnedFractionChange::Inc(*pos - sub),
+                Ordering::Equal => OwnedFractionChange::Zero,
+                Ordering::Greater => OwnedFractionChange::Dec(sub - *pos),
+            },
         };
     }
 
     fn merge_received(&mut self, add: Self::State) {
         *self = match self {
             OwnedFractionChange::Inc(pos) => OwnedFractionChange::Inc(*pos + add),
             OwnedFractionChange::Zero => OwnedFractionChange::Inc(add),
-            OwnedFractionChange::Dec(neg) if *neg > add => OwnedFractionChange::Dec(*neg - add),
-            OwnedFractionChange::Dec(neg) if *neg == add => OwnedFractionChange::Zero,
-            OwnedFractionChange::Dec(neg) if *neg < add => OwnedFractionChange::Inc(add - *neg),
-            OwnedFractionChange::Dec(_) => unreachable!(),
+            OwnedFractionChange::Dec(neg) => match add.cmp(neg) {
+                Ordering::Greater => OwnedFractionChange::Dec(*neg - add),
```

Comment:
This must be less, not greater - the order of comparison is reversed in the match